link.go

```diff
@@ -71,41 +71,163 @@ type link struct {
 	msg                   Message             // current message being decoded
 }
 
-func newLink(s *Session, r *Receiver, opts []LinkOption) (*link, error) {
+// newSendingLinkd creates a new sending link and attaches it to the session
+func newSendingLink(target string, s *Session, opts *SenderOptions) (*link, error) {
 	l := &link{
-		Key:                      linkKey{randString(40), encoding.Role(r != nil)},
+		Key:                      linkKey{randString(40), encoding.RoleSender},
 		Session:                  s,
-		receiver:                 r,
 		close:                    make(chan struct{}),
 		Detached:                 make(chan struct{}),
 		ReceiverReady:            make(chan struct{}, 1),
 		detachOnDispositionError: true,
+		Target:                   &frames.Target{Address: target},
+		Source:                   new(frames.Source),
+	}
+
+	if opts == nil {
+		return l, nil
 	}
 
-	// configure options
-	for _, o := range opts {
-		err := o(l)
-		if err != nil {
+	for _, v := range opts.Capabilities {
+		l.Source.Capabilities = append(l.Source.Capabilities, encoding.Symbol(v))
+	}
+	if opts.Durability > DurabilityUnsettledState {
+		return nil, fmt.Errorf("invalid Durability %d", opts.Durability)
+	}
+	l.Source.Durable = opts.Durability
+	if opts.DynamicAddress {
+		l.Target.Address = ""
+		l.dynamicAddr = opts.DynamicAddress
+	}
+	if opts.ExpiryPolicy != "" {
+		if err := encoding.ValidateExpiryPolicy(opts.ExpiryPolicy); err != nil {
 			return nil, err
 		}
+		l.Source.ExpiryPolicy = opts.ExpiryPolicy
+	}
+	l.Source.Timeout = opts.ExpiryTimeout
+	l.detachOnDispositionError = !opts.IgnoreDispositionErrors
+	if opts.Name != "" {
+		l.Key.name = opts.Name
+	}
+	if opts.Properties != nil {
+		l.properties = make(map[encoding.Symbol]interface{})
+		for k, v := range opts.Properties {
+			if k == "" {
+				return nil, errors.New("link property key must not be empty")
+			}
+			l.properties[encoding.Symbol(k)] = v
+		}
+	}
+	if opts.RequestedReceiverSettleMode != nil {
+		if rsm := *opts.RequestedReceiverSettleMode; rsm > ModeSecond {
+			return nil, fmt.Errorf("invalid RequestedReceiverSettleMode %d", rsm)
+		}
+		l.ReceiverSettleMode = opts.RequestedReceiverSettleMode
+	}
+	if opts.SettlementMode != nil {
+		if ssm := *opts.SettlementMode; ssm > ModeMixed {
+			return nil, fmt.Errorf("invalid SettlementMode %d", ssm)
+		}
+		l.SenderSettleMode = opts.SettlementMode
+	}
+	l.Source.Address = opts.SourceAddress
+	return l, nil
+}
+
+func newReceivingLink(source string, s *Session, r *Receiver, opts *ReceiverOptions) (*link, error) {
+	l := &link{
+		Key:                      linkKey{randString(40), encoding.RoleReceiver},
+		Session:                  s,
+		receiver:                 r,
+		close:                    make(chan struct{}),
+		Detached:                 make(chan struct{}),
+		ReceiverReady:            make(chan struct{}, 1),
+		detachOnDispositionError: true,
```

Comment:
Looking at the code that uses this:

```go
if l.detachOnDispositionError && (l.receiver == nil && (l.ReceiverSettleMode == nil || *l.ReceiverSettleMode == ModeFirst)) {
```

It looks like this is a sender only option, not a receiver option. I suppose we could just leave it as 'false' since we know we're a receiver and then you can simplify the check in `detachOnRejectDisp`.

Reply:
Good point about this being sender-only.  I will remove the `IgnoreDispositionErrors` field from `ReceiverOptions`.

Reply:
Fixed